cmd/load-generator/wfe/state.go

```diff
@@ -0,0 +1,431 @@
+	if err != nil {
+		return nil, err
+	}
+	return &State{
+		rMu:           new(sync.RWMutex),
+		nMu:           new(sync.RWMutex),
+		challRPCAddr:  rpcAddr,
+		client:        client,
+		apiBase:       apiBase,
+		throughput:    int64(rate),
+		certKey:       certKey,
+		domainBase:    domainBase,
+		callLatency:   latencyFile,
+		runtime:       runtime,
+		termsURL:      termsURL,
+		wg:            new(sync.WaitGroup),
+		realIP:        realIP,
+		runPlan:       runPlan,
+		maxRegs:       maxRegs,
+		warmupWorkers: warmupWorkers,
+		warmupRegs:    warmupRegs,
+	}, nil
+}
+
+func (s *State) executePlan() {
+	for _, p := range s.runPlan {
+		atomic.StoreInt64(&s.throughput, p.Rate)
+		fmt.Printf("Set base action rate to %d/s for %s\n", p.Rate, p.For)
+		time.Sleep(p.For)
+	}
+}
+
+func (s *State) warmup() {
+	fmt.Printf("Beginning warmup, generating ~%d registrations with %d workers\n", s.warmupRegs, s.warmupWorkers)
+	wg := new(sync.WaitGroup)
+	for i := 0; i < s.warmupWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			for {
+				s.rMu.RLock()
+				if len(s.regs) >= s.warmupRegs {
+					s.rMu.RUnlock()
+					break
+				}
+				s.rMu.RUnlock()
+
+				s.newRegistration(nil)
+			}
+			wg.Done()
+		}()
+	}
+	stopProg := make(chan bool, 1)
+	go func() {
+		var last string
+		for _ = range time.Tick(1 * time.Second) {
+			select {
+			case <-stopProg:
+				return
+			default:
+				if last != "" {
+					fmt.Fprintf(os.Stdout, strings.Repeat("\b", len(last)))
+				}
+				s.rMu.RLock()
+				last = fmt.Sprintf("%d/%d registrations generated", len(s.regs), s.warmupRegs)
+				fmt.Fprint(os.Stdout, last)
+				s.rMu.RUnlock()
+			}
+		}
+	}()
+	wg.Wait()
+	stopProg <- true
+	fmt.Println("\nFinished warming up")
+}
+
+// Run runs the WFE load-generator for either the specified runtime/rate or the execution plan
+func (s *State) Run(binName string, dontRunChallSrv bool, httpOneAddr string) error {
+	// If warmup, warmup
+	if s.warmupRegs > len(s.regs) {
+		s.warmup()
+	}
+
+	// Start chall server process
+	if !dontRunChallSrv {
+		cmd := exec.Command(binName, "chall-srv", "--rpcAddr="+s.challRPCAddr, "--httpOneAddr="+httpOneAddr)
+		err := cmd.Start()
+		if err != nil {
+			return err
+		}
+		s.challSrvProc = cmd.Process
+	}
+
+	// If there is a run plan execute it
+	if len(s.runPlan) > 0 {
+		go s.executePlan()
+	}
+
+	// Run sending loop
+	stop := make(chan bool, 1)
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		for {
+			select {
+			case <-stop:
+				return
+			default:
+				s.wg.Add(1)
+				go s.sendCall()
+				time.Sleep(time.Duration(time.Second.Nanoseconds() / atomic.LoadInt64(&s.throughput)))
+			}
+		}
+	}()
+
+	select {
+	case <-time.After(s.runtime):
+		fmt.Println("SLEEP END")
+	case sig := <-sigs:
+		fmt.Printf("SIG CAUGHT [%s], ENDING\n", sig.String())
+	}
+	stop <- true
+	fmt.Println("SENT STOP")
+	s.wg.Wait()
+	fmt.Println("KILLING CHALL SERVER")
+	err := s.challSrvProc.Kill()
```

Comment:
If the user provided `--dontRunChallSrv` then the load generator should **not** try to kill the challenge server when it is done. Otherwise a panic is produced:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x483569]

goroutine 1 [running]:
panic(0x979940, 0xc82000e120)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
os.(*Process).signal(0x0, 0x7efde554d078, 0xc82000e268, 0x0, 0x0)
	/usr/local/go/src/os/exec_unix.go:40 +0x39
os.(*Process).Signal(0x0, 0x7efde554d078, 0xc82000e268, 0x0, 0x0)
	/usr/local/go/src/os/doc.go:55 +0x41
os.(*Process).kill(0x0, 0x0, 0x0)
	/usr/local/go/src/os/exec_posix.go:53 +0x45
os.(*Process).Kill(0x0, 0x0, 0x0)
	/usr/local/go/src/os/doc.go:40 +0x2d
github.com/letsencrypt/boulder/cmd/load-generator/wfe.(*State).Run(0xc82016a200, 0x7ffece4aa30f, 0x3b, 0xc820160701, 0xc8201607a0, 0xe, 0x0, 0x0)
	/home/daniel/go/src/github.com/letsencrypt/boulder/cmd/load-generator/wfe/state.go:284 +0x8ad
main.main.func1(0xc8201c80c0)
	/home/daniel/go/src/github.com/letsencrypt/boulder/cmd/load-generator/main.go:179 +0x72b
github.com/letsencrypt/boulder/vendor/github.com/codegangsta/cli.Command.Run(0xa357a8, 0x3, 0x0, 0x0, 0xa3da40, 0xd, 0x0, 0x0, 0x0, 0x0, ...)
	/home/daniel/go/src/github.com/letsencrypt/boulder/vendor/github.com/codegangsta/cli/command.go:113 +0x1087
github.com/letsencrypt/boulder/vendor/github.com/codegangsta/cli.(*App).Run(0xc82016c820, 0xc82000a080, 0x8, 0x8, 0x0, 0x0)
	/home/daniel/go/src/github.com/letsencrypt/boulder/vendor/github.com/codegangsta/cli/app.go:156 +0xbc0
main.main()
	/home/daniel/go/src/github.com/letsencrypt/boulder/cmd/load-generator/main.go:270 +0x21b3
exit status 2
```